x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/retriever/PinnedRetrieverBuilder.java

```diff
@@ -0,0 +1,266 @@
+        List<SortBuilder<?>> sorts = source.sorts();
+        if (sorts == null || sorts.isEmpty()) {
+            return;
+        }
+        for (SortBuilder<?> sort : sorts) {
+            if (sort instanceof ScoreSortBuilder) {
+                continue;
+            }
+            if (sort instanceof FieldSortBuilder) {
+                FieldSortBuilder fieldSort = (FieldSortBuilder) sort;
+                if (ShardDocSortField.NAME.equals(fieldSort.getFieldName())) {
+                    continue;
+                }
+            }
+            throw new IllegalArgumentException(
+                "[" + NAME + "] retriever only supports sorting by score, invalid sort criterion: " + sort.toString()
+            );
+        }
+    }
+
+    public PinnedRetrieverBuilder(List<String> ids, List<SpecifiedDocument> docs, RetrieverBuilder retrieverBuilder, int rankWindowSize) {
+        super(new ArrayList<>(), rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids != null ? ids : new ArrayList<>();
+        this.docs = docs != null ? docs : new ArrayList<>();
+        addChild(new PinnedRetrieverBuilderWrapper(retrieverBuilder));
+    }
+
+    public PinnedRetrieverBuilder(
+        List<String> ids,
+        List<SpecifiedDocument> docs,
+        List<RetrieverSource> retrieverSource,
+        int rankWindowSize,
+        String retrieverName,
+        List<QueryBuilder> preFilterQueryBuilders
+    ) {
+        super(retrieverSource, rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids;
+        this.docs = docs;
+        this.retrieverName = retrieverName;
+        this.preFilterQueryBuilders = preFilterQueryBuilders;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    public int rankWindowSize() {
+        return rankWindowSize;
+    }
+
+    /**
+     * Creates a PinnedQueryBuilder with the appropriate pinned documents.
+     * Prioritizes docs over ids if both are present.
+     *
+     * @param baseQuery the base query to pin documents to
+     * @return a PinnedQueryBuilder
+     * @throws IllegalArgumentException if baseQuery is null
+     */
+    private QueryBuilder createPinnedQuery(QueryBuilder baseQuery) {
+        Objects.requireNonNull(baseQuery, "Underlying query cannot be null for pinned retriever");
+
+        if (docs.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, docs.toArray(new SpecifiedDocument[0]));
+        }
+        return new PinnedQueryBuilder(baseQuery, ids.toArray(new String[0]));
+    }
+
+    @Override
+    protected SearchSourceBuilder finalizeSourceBuilder(SearchSourceBuilder source) {
+        validateSort(source);
+        source.query(createPinnedQuery(source.query()));
+        return source;
+    }
+
+    @Override
+    public void doToXContent(XContentBuilder builder, Params params) throws IOException {
+        if (ids != null && ids.isEmpty() == false) {
```

Comment:
This would read a bit cleaner if you'd applied my previous PR feedback about not initializing this to empty list.